test/e2e/app/main.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/tendermint/tendermint/abci/server"
+	"github.com/tendermint/tendermint/libs/log"
+	tmnet "github.com/tendermint/tendermint/libs/net"
+	"github.com/tendermint/tendermint/privval"
+)
+
+var logger = log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+
+// main is the binary entrypoint.
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Printf("Usage: %v <configfile>", os.Args[0])
+		return
+	}
+	configFile := ""
+	if len(os.Args) == 2 {
+		configFile = os.Args[1]
+	}
+
+	if err := run(configFile); err != nil {
+		logger.Error(err.Error())
+		os.Exit(1)
+	}
+}
+
+// run runs the application - basically like main() with error handling.
+func run(configFile string) error {
+	// Set up application
+	cfg, err := LoadConfig(configFile)
+	if err != nil {
+		return err
+	}
+	app, err := NewApplication(cfg)
+	if err != nil {
+		return err
+	}
+
+	// Start ABCI server
+	protocol := "socket"
+	if cfg.GRPC {
+		protocol = "grpc"
+	}
```

Comment:
Should this allow setting in process as well? maybe a follow-up item?

Reply:
Yeah, I was considering it, but it seemed like it would get pretty fiddly and I wasn't sure how useful it would be. Wouldn't we basically need to build a completely different binary to launch instead of Tendermint, and do all of the config handling and such separately?

Reply:
Not necessarily. It could be the same way but yes, the binary would have to be built separately. The tradeoff in testing by using socket or unix is mainly performance I would say. 

An alternative approach could be `--proxy_app` and just adding your new app to the switch here: https://github.com/tendermint/tendermint/blob/2d167aefcff234de5f85b283aacfd9a8b04c7548/proxy/client.go#L72-L72

Reply:
> Well, we need to exercise the socket and grpc protocol implementations anyway, so it's not like we can drop them even if we implement in-app support - it would be an additional thing to test. 

Agree, meant as a third option 🙈

Reply:
Done. It's a bit messy, but it works.